src/mlpack/methods/bayesian_ridge/bayesian_ridge.cpp

```diff
@@ -0,0 +1,262 @@
+  dataOffset = arma::zeros<arma::colvec>(data.n_rows);
+  dataScale = arma::ones<arma::colvec>(data.n_rows);
+  responsesOffset = 0.0;
+
+  if (centerData)
+  {
+    dataOffset = mean(data, 1);
+    responsesOffset = mean(responses);
+  }
+
+  if (scaleData)
+    dataScale = stddev(data, 0, 1);
+
+  // Copy data and response before the processing.
+  dataProc = data;
+  // Center the data.
+  dataProc.each_col() -= dataOffset;
+  // Scale the data.
+  dataProc.each_col() /= dataScale;
+  // Center the responses.
+  responsesProc = responses - responsesOffset;
+
+  return responsesOffset;
+}
+
+
+// Copy construcor.
+BayesianRidge::BayesianRidge(const BayesianRidge& other):
+  centerData(other.centerData),
+  scaleData(other.scaleData),
+  dataOffset(other.dataOffset),
+  dataScale(other.dataScale),
+  responsesOffset(other.responsesOffset),
+  alpha(other.alpha),
+  beta(other.beta),
+  gamma(other.gamma),
+  omega(other.omega),
+  matCovariance(other.matCovariance)
+{/* All is done */}
+
+// Move constructor.
+BayesianRidge::BayesianRidge(BayesianRidge&& other):
+  centerData(other.centerData),
+  scaleData(other.scaleData),
+  dataOffset(std::move(other.dataOffset)),
+  dataScale(std::move(other.dataScale)),
+  responsesOffset(other.responsesOffset),
+  alpha(other.alpha),
+  beta(other.beta),
+  gamma(other.gamma),
+  omega(std::move(other.omega)),
+  matCovariance(std::move(other.matCovariance))
+{
+  // Clear the other object.
+  if (this != &other)
+  {
+    other.centerData = false;
+    other.scaleData = false;
+    other.dataOffset.reset();
+    other.dataScale.reset();
+    other.responsesOffset = 0.0;
+    other.alpha = 0.0;
+    other.gamma = 0.0;
+    other.beta = 0.0;
+    other.omega.reset();
+    other.matCovariance.reset();
+  }
+}
+
+BayesianRidge& BayesianRidge::operator=(const BayesianRidge& other)
+{
+  if (this == &other)
+    return *this;
+
+  centerData = other.centerData;
+  scaleData = other.scaleData;
+  dataOffset = other.dataOffset;
+  dataScale = other.dataScale;
+  responsesOffset = other.responsesOffset;
+  alpha = other.alpha;
+  gamma = other.gamma;
+  beta = other.beta;
+  omega = other.omega;
+  matCovariance = other.matCovariance;
+  return *this;
+}
+
+BayesianRidge& BayesianRidge::operator=(BayesianRidge&& other)
+{
+  if (this != &other)
+  {
+    centerData = other.centerData;
+    scaleData = other.scaleData;
+    dataOffset = other.dataOffset;
+    dataScale = other.dataScale;
+    responsesOffset = other.responsesOffset;
+    alpha = other.alpha;
+    gamma = other.gamma;
+    beta = other.beta;
+    omega = other.omega;
+    matCovariance = other.matCovariance;
+
+    // Clear the other object.
+    other.centerData = false;
+    other.scaleData = false;
+    other.dataOffset.reset();
+    other.dataScale.reset();
+    other.responsesOffset = 0.0;
+    other.alpha = 0.0;
+    other.gamma = 0.0;
+    other.beta = 0.0;
+    other.omega.reset();
+    other.matCovariance.reset();
+  }
+  return *this;
+}
```

Comment:
Why can't we just use the default versions of these methods?  The only thing not copied is the number of iterations and the tolerance, and it seems like those would be things we want to copy anyway.

Reply:
Do you mean that the compiler generates the move constructor automatically ? I was not sure so I wrote them by default after reading this in the design guidelines: 
"mlpack objects should provide copy and move constructors; the copy constructor should have warnings in its documentation if the copy may be memory-intensive."
If they are useless let's suppress them !
The number of iterations and the tolerance have been added later thanks to the @zoq advises. I forgot to report them in those functions.

Reply:
Ahh, okay, it sounds like maybe we should update the design guidelines.  I think, if I remember right, that if I wrote those (which... I think I did?) it was a very long time ago and I wasn't considering default copy and move operators.  In fact... yes... I think I remember... when I started maintaining mlpack in early 2010, the library designers at that time would disable copy constructors by marking them private.  This was a way to avoid expensive copying of datasets and data structures.  So, that note in the design guidelines is actually just a response to that (anti?)pattern of private copy constructors.

I went ahead and updated the design guidelines to be a bit more clear.  Thank you for pointing this out!  It was a fun trip down memory lane. :)

Anyway, it looks like the default versions of all these methods should be sufficient here, so, unless I overlooked something, I guess we can remove these. :+1: